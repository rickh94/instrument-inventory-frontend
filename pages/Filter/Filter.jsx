import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { RootPaper, LoadingHeader } from '../../components'
import {
  FormControl,
  FormControlLabel,
  Checkbox,
  FormGroup,
  Button
} from '@material-ui/core'
import { Fields } from '../../components'
import { lastButton } from '../../globalStyles'
import { withStyles } from '@material-ui/styles';

const styles = {
  lastButton
}

class Filter extends Component {
  constructor(props) {
    super(props)

    this.state = {
      isLoading: false,
      type: '',
      size: '',
      location: '',
      notAssigned: false
    }
  }

  onSubmit = () => {}

  handleChange = name => event => {
    this.setState({ [name]: event.target.value })
  }

  handleCheck = name => event => {
    this.setState({ [name]: event.target.checked })
  }

  validateForm = () => {
    return (this.state.type || this.state.size || this.state.location)
  }

  render() {
    const { classes } = this.props
    const { type, size, location, notAssigned } = this.state
    return (
      <RootPaper>
        <LoadingHeader isLoading={this.state.isLoading} title="Filter Instruments" />
        <form onSubmit={this.onSubmit}>
          <Fields.InstrumentTypeSelect
            error={false}
            value={type}
            onChange={this.handleChange('type')}
          />
          <Fields.InstrumentSizeSelect
            error={false}
            value={size}
            onChange={this.handleChange('size')}
            instrumentType={type}
          />
          <Fields.LocationSelect
            error={false}
            value={location}
            onChange={this.handleChange('location')}
          />
          <FormControlLabel
            control={
              <Checkbox
                checked={notAssigned}
                onChange={this.handleCheck('notAssigned')}
              />
            }
            label="Show only unassigned instruments"
          />
          <FormGroup row>
            <Button onClick={this.clearForm} className={classes.lastButton}>
              Clear
            </Button>
            <Button type="submit" color="primary" disabled={!this.validateForm()}>
              Submit
            </Button>
          </FormGroup>
        </form>
      </RootPaper>
    )
  }
}

Filter.propTypes = {
  history: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(Filter)
