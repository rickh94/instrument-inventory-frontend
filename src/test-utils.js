/* eslint-disable react/prop-types */
import React from 'react'
import { MuiThemeProvider, createMuiTheme, indigo } from '@material-ui/core'
import { HelpersContext, SchemaContext } from './contexts'
import { MemoryRouter } from 'react-router-dom'
import { render } from '@testing-library/react'

const theme = createMuiTheme({
  palette: {
    primary: indigo,
    secondary: {
      main: '#0A81D1',
    },
  },
})

export const showAlert = jest.fn()
export const setSearchResults = jest.fn()
export const setFilterResults = jest.fn()
export const searchResults = jest.fn()
export const filterResults = jest.fn()

export const TestTheme = ({ children }) => (
  <MuiThemeProvider theme={theme}>{children}</MuiThemeProvider>
)

export const TestHelpers = ({ children }) => (
  <HelpersContext.Provider
    value={{
      showAlert,
      setSearchResults,
      setFilterResults,
      searchResults,
      filterResults,
    }}
  >
    {children}
  </HelpersContext.Provider>
)

export const TestRouter = ({ children }) => <MemoryRouter>{children}</MemoryRouter>

export const TestSchema = ({ children, schema = { components: { schemas: {} } } }) => (
  <SchemaContext.Provider value={schema}>{children}</SchemaContext.Provider>
)

export const TestEverything = ({ children }) => (
  <TestRouter>
    <TestTheme>
      <TestHelpers>{children}</TestHelpers>
    </TestTheme>
  </TestRouter>
)

// eslint-disable-next-line no-undef
export const flushPromises = () => new Promise(setImmediate)

const customRender = (ui, options, schema = { components: { schemas: {} } }) =>
  render(<TestSchema schema={schema}>{ui}</TestSchema>, {
    wrapper: TestEverything,
    ...options,
  })

export * from '@testing-library/react'

export { customRender as render }
