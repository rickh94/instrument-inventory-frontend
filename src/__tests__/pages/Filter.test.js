import React from 'react'
import Filter from '../../pages/Filter'
import { render, cleanup } from '../../test-utils'

const schema = {
  openapi: '3.0.2',
  components: {
    schemas: {
      InstrumentFilter: {
        title: 'Instrument Filter',
        type: 'object',
        properties: {
          location: {
            title: 'Location',
            enum: ['Office', 'Storage'],
          },
          notAssigned: {
            default: false,
            type: 'boolean',
            title: 'Search only unassigned instruments',
          },
          size: {
            title: 'Size',
            enum: ['4/4', '3/4', '1/2', '1/4'],
          },
          type: {
            enum: ['Violin', 'Viola', 'Cello', 'Bass', 'Violin strung as viola'],
            title: 'Instrument Type',
          },
        },
      },
    },
  },
}

afterEach(cleanup)

describe('<Filter />', () => {
  test('matches snapshot', () => {
    const { container } = render(<Filter history={{ push: jest.fn() }} />, {}, schema)
    expect(container).toMatchSnapshot()
  })

  test('renders correct fields', () => {
    const { queryByText } = render(<Filter history={{ push: jest.fn() }} />, {}, schema)
    expect(queryByText('Location')).toBeTruthy()
    expect(queryByText('Search only unassigned instruments')).toBeTruthy()
    expect(queryByText('Size')).toBeTruthy()
    expect(queryByText('Instrument Type')).toBeTruthy()
  })
})
