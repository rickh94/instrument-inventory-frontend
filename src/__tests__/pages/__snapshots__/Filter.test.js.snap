// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> matches snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 WithStyles(ForwardRef(Paper))-root-1 MuiPaper-rounded"
  >
    <div
      class="makeStyles-heading-29"
      data-testid="loading-header"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 MuiTypography-colorInherit"
      >
        Filter Instruments
      </h5>
    </div>
    <form
      data-testid="schema-form"
      novalidate=""
    >
      <div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
            data-shrink="false"
          >
            Location
          </label>
          <div
            appearance="checkbox"
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputSelect MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
              id="select-location"
              role="button"
              tabindex="0"
            >
              <span>
                ​
              </span>
            </div>
            <input
              id="uniforms-0000-0001"
              name="location"
              type="hidden"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <fieldset
          class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        >
          <div
            class="MuiFormGroup-root"
          >
            <label
              class="MuiFormControlLabel-root"
            >
              <span
                aria-disabled="false"
                class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-151 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                  <input
                    class="PrivateSwitchBase-input-154"
                    data-indeterminate="false"
                    id="uniforms-0000-0003"
                    name="notAssigned"
                    type="checkbox"
                    value="notAssigned"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              >
                Search only unassigned instruments
              </span>
            </label>
          </div>
        </fieldset>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
            data-shrink="false"
          >
            Size
          </label>
          <div
            appearance="checkbox"
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputSelect MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
              id="select-size"
              role="button"
              tabindex="0"
            >
              <span>
                ​
              </span>
            </div>
            <input
              id="uniforms-0000-0005"
              name="size"
              type="hidden"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense"
            data-shrink="false"
          >
            Instrument Type
          </label>
          <div
            appearance="checkbox"
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputSelect MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
              id="select-type"
              role="button"
              tabindex="0"
            >
              <span>
                ​
              </span>
            </div>
            <input
              id="uniforms-0000-0007"
              name="type"
              type="hidden"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="MuiFormGroup-root MuiFormGroup-row"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-button-61 makeStyles-firstButton-63 makeStyles-lastButton-62 MuiButton-text MuiButton-textPrimary"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Submit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </div>
</div>
`;
